dbms/src/Interpreters/Expand.cpp

```diff
@@ -0,0 +1,260 @@
+    {
+        current_offset += replicate_times_for_one_row;
+        (*offsets_to_replicate)[i] = current_offset;
+
+        // in the same loop, to fill the grouping id.
+        for (UInt64 j = 0; j < replicate_times_for_one_row; j++)
+        {
+            // start from 1.
+            Field grouping_id = j + 1;
+            added_grouping_id_column[0]->insert(grouping_id);
+        }
+    }
+    // todo: for some column overlapping in different grouping set, we should copy the overlapped column as a new column
+    //  and the upper layer OP's computation should be shifted and based on the new one's id. Need a plan side control.
+
+    // replicate the original block rows.
+    size_t existing_columns = block.columns();
+
+    if (offsets_to_replicate)
+    {
+        for (size_t i = 0; i < existing_columns; ++i)
+        {
+            // expand the origin const column, since it may be filled with null value when expanding.
+            if (block.safeGetByPosition(i).column->isColumnConst())
+                block.safeGetByPosition(i).column = block.safeGetByPosition(i).column->convertToFullColumnIfConst();
+
+            // for every existing column, if the column is a grouping set column, make it nullable.
+            if (isInGroupSetColumn(block.safeGetByPosition(i).name) && !block.safeGetByPosition(i).column->isColumnNullable())
+            {
+                convertColumnToNullable(block.getByPosition(i));
+            }
+            if (!offsets_to_replicate->empty())
+                // replicate it.
+                block.safeGetByPosition(i).column = block.safeGetByPosition(i).column->replicate(*offsets_to_replicate);
+        }
+    }
+
+
+    // after replication, it just copied the same row for N times, we still need to fill corresponding Field with null value.
+    for (size_t grouping_offset = 0; grouping_offset < replicate_times_for_one_row; grouping_offset++)
+    {
+        auto grouping_columns = getGroupSetColumnNamesByOffset(grouping_offset);
+        // for every grouping col, get the mutated one of them.
+        for (const auto & grouping_col : grouping_columns)
+        {
+            assert(block.getByName(grouping_col).column->isColumnNullable());
+
+            const auto * nullable_column = typeid_cast<const ColumnNullable *>(block.getByName(grouping_col).column.get());
+            auto origin_size = nullable_column->size();
+            // clone the nested column.
+            MutableColumnPtr new_nested_col = nullable_column->getNestedColumn().cloneResized(origin_size);
+            // just get mutable new null map.
+            auto new_null_map = ColumnUInt8::create();
+            new_null_map->getData().resize(origin_size);
+            memcpy(new_null_map->getData().data(), nullable_column->getNullMapData().data(), origin_size * sizeof(nullable_column->getNullMapData()[0]));
+
+            auto cloned_one = ColumnNullable::create(std::move(new_nested_col), std::move(new_null_map));
```

Comment:
Looks like just want to copy nullable_column to cloned_one. These codes are duplicated. Maybe just:
```
auto cloned_one = ColumnNullable::create(nullable_column.getNestedColumnPtr(), nullable_column.getNullMapColumnPtr());
```

Reply:
actually I need a new clone one and return mutablePtr here, which can support me to assign data directly later and has not influence to the old one.

after trails, it seems the your cloned_one here will call a const type getNullMapData
```
assignment statement: cloned_one->getNullMapData().data()[computed_offset] = 1;
                                       ***

my:     NullMap & getNullMapData() { return getNullMapColumn().getData(); }
you:    const NullMap & getNullMapData() const { return getNullMapColumn().getData(); }
```

const type func are not allow to direct data assignment here >.<!!!